File: Data Structures & Algorithms/anagram-groups/submission-0.py
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        # input is array of strings
        # output is sublist of strings with anagrams in their own lists inside it
        # how to get there = group same length strings together, then make hashmap for each string with letter type and count
        #   then check which ones are equal, group equal ones together

        # whats this?
        res = defaultdict(list)

        # go through each string in strs
        for s in strs:
            # create letter array
            count = [0] * 26
            # go through each char in that string and add 1 to the array index that corresponds to the letter
            for c in s:
                count[ord(c) - ord('a')] += 1 # find the index for that letter c, count[88(b) - 87(a)]  = count[1] 
                                                # then add 1 to the value for that letters index, count[1] += 1
            # lists cant be keys, so need to change it into a tuple
            res[tuple(count)].append(s)
        return res.values()

File: Data Structures & Algorithms/duplicate-integer/submission-0.java
class Solution {
    public boolean hasDuplicate(int[] nums) {
        // input = int array 
        // output = true if multiple values
        //         false if all values are unique
        // how to get there = make a hash map, set the values in num array to be the key with 0 as the value
        // check if hash map already contains the key, if so return true
        // if i hit no duplicate "key" i set the value to 0
        // add key/value to hash map as i go with getOrDefault((array[val], 0)+1)
        

        HashMap<Integer, Integer> map = new HashMap<>();

        // loop through array and check or add value to hashmap
        for (int i : nums) {
            if (map.containsKey(i)) return true;
            else {

                map.put(i, 0);
            }
        }
        return false;
    }
}
